### PR TITLE
Overflow prevention

### PR DESCRIPTION
**Pull request by paulidale**
Addressing strcpy, strcat and sprintf issues in `crypto/bn/bn_print.c`, `crypto/mem_dbg.c` and `crypto/pem/pem_lib.c`

Also replace `return (x)` with `return x`

### CODE REVIEW

File: crypto/bn/bn_print.c [hunk 1]
```diff
@@ -67,9 +67,10 @@ char *BN_bn2dec(const BIGNUM *a)
      */
     i = BN_num_bits(a) * 3;
     num = (i / 10 + i / 1000 + 1) + 1;
+    tbytes = num + 3;
```
Comment (richsalz) [inline crypto/bn/bn_print.c]: why +3?  A comment here would help.

File: crypto/pem/pem_lib.c [hunk 2]
```diff
@@ -71,6 +71,7 @@ int PEM_def_callback(char *buf, int num, int w, void *key)
 void PEM_proc_type(char *buf, int type)
 {
     const char *str;
+    char *p = strchr(buf, '\0');
```
Comment (richsalz) [inline crypto/pem/pem_lib.c]: is this anything other than buf+strlen(buf) ?

File: crypto/pem/pem_lib.c [hunk 3]
```diff
@@ -81,27 +82,29 @@ void PEM_proc_type(char *buf, int type)
...
-    strcat(buf, type);
-    strcat(buf, ",");
-    j = strlen(buf);
-    for (i = 0; i < len; i++) {
-        buf[j + i * 2] = map[(str[i] >> 4) & 0x0f];
-        buf[j + i * 2 + 1] = map[(str[i]) & 0x0f];
-    }
-    buf[j + i * 2] = '\n';
-    buf[j + i * 2 + 1] = '\0';
+    char *p = strchr(buf, '\0');
```
Comment (richsalz) [inline crypto/pem/pem_lib.c]: same strlen comment.

File: crypto/pem/pem_lib.c [hunk 4]
```diff
@@ -81,27 +82,29 @@ void PEM_proc_type(char *buf, int type)
...
+    int j = PEM_BUFSIZE - (p - buf), n;
```
Comment (richsalz) [inline crypto/pem/pem_lib.c]: int  cast on the pointer subtraction

File: crypto/pem/pem_lib.c [hunk 5]
```diff
@@ -81,27 +82,29 @@ void PEM_proc_type(char *buf, int type)
...
+        p += n;
+        for (i = 0; i < len; i++) {
+            n = BIO_snprintf(p, j, "%02X", 0xff & str[i]);
+            if (n <= 0)
+                return;
+            j -= n;
+            p += n;
+        }
+        if (j > 1)
+            memcpy(p, "\n", sizeof("\n"));
```
Comment (richsalz) [inline crypto/pem/pem_lib.c]: strcpy(p, "\n") seems more clear.

File: crypto/pem/pem_lib.c [hunk 6]
```diff
@@ -570,14 +573,14 @@ static int load_iv(char **fromp, unsigned char *to, int num)
         v = OPENSSL_hexchar2int(*from);
         if (v < 0) {
             PEMerr(PEM_F_LOAD_IV, PEM_R_BAD_IV_CHARS);
-            return (0);
+            return 0;
         }
         from++;
         to[i / 2] |= v << (long)((!(i & 1)) * 4);
```
Comment (richsalz) [inline crypto/pem/pem_lib.c]: side note: whoever wrote that expression should be smacked.

File: crypto/pem/pem_lib.c (hunk 2 above)
Comment (paulidale) [inline crypto/pem/pem_lib.c]: Nope, they do the same thing.

Want these changed?

**Comment by richsalz** [comment]
Yes, please change them; strlen is more obvious.

File: crypto/bn/bn_print.c (hunk 1 above)
Comment (paulidale) [inline crypto/bn/bn_print.c]: This was in the original code in the malloc which was comment free.

Negative sign and null terminator are two.  I've no idea why 3.
The derivation of the bound on num in the comment above looks correct.

**Comment by richsalz** [comment]
Okay fine

File: crypto/pem/pem_lib.c (hunk 6 above)
Comment (paulidale) [inline crypto/pem/pem_lib.c]: Agree.  I didn't notice it.

**Comment by paulidale** [comment]
Changes are in.

**Pull request closed without merging**